apptests/test_generators.py

```diff
@@ -0,0 +1,326 @@
+                        "width": 120,
+                        "height": 90,
+                    }
+                },
+                "channelTitle": "SMS Frankfurt Group Travel",
+                "publishTime": "2016-12-21T11:09:18Z",
+            },
+        }
+    ],
+}
+parsed_example = [
+    {
+        "videoId": "YazZwd48ws0",
+        "channelId": "UCWlvkaA27BCrcYG6gA0K8UA",
+        "title": "We create your Bucket List Trips and Events | Discover the Baltics with SMS Frankfurt Group Travel",
+        "channelTitle": "SMS Frankfurt Group Travel",
+        "publishTime": "2016-12-21T11:09:18Z",
+    }
+]
+
+
+search_request = {
+    "args": {
+        "maxResults": 10,
+        "pageToken": None,
+        "q": "Travel & Events",
+        "regionCode": "US",
+        "relevanceLanguage": "en",
+        "topicId": None,
+        "videoCategoryId": "19",
+        "videoDuration": "medium",
+        "videoLicense": "creativeCommon",
+    }
+}
+
+
+results_example = {
+    "args": {
+        "maxResults": 10,
+        "q": "Travel & Events",
+        "regionCode": "US",
+        "relevanceLanguage": "en",
+        "videoCategoryId": "19",
+        "videoDuration": "medium",
+        "videoLicense": "creativeCommon",
+        "pageToken": "CAEQAA",
+    }
+}
+
+
+class Test_videos_details_request(unittest.TestCase):
+    def setUp(self) -> None:
+        self.expected_keys = ["kind", "etag", "items", "pageInfo"]
+        self.valid_ids = ["gI8VlFK5Jp4", "YazZwd48ws0"]
+        return super().setUp()
+
+    def test_returned_type(self):
+        self.assertIsInstance(gen.videos_details_request([]), dict)
+
+    def test_empty_list(self):
+        self.assertEqual(gen.videos_details_request([]), empty_response_details)
+
+    def test_invalid_ids(self):
+        self.assertEqual(
+            gen.videos_details_request(["invalid_id"]), empty_response_details
+        )
+
+    def test_valid_ids(self):
+        self.assertNotEqual(
+            gen.videos_details_request(self.valid_ids),
+            empty_response_details,
+        )
+
+    def test_valid_ids_response_keys(self):
+        response = gen.videos_details_request(self.valid_ids)
+
+        for key in self.expected_keys:
+            with self.subTest(key=key):
+                self.assertIn(key, response)
+
+
+class Test_categories_request(unittest.TestCase):
+    # should return info about 400 response?
+    # def test_returned_type(self):
+    #     self.assertIsInstance(gen.categories_request("", ""), dict)
+
+    # def test_invalid_params(self):
+    #     self.assertIsNotNone(gen.categories_request("invalid", "invalid").get("error"))
+
+    # def test_empty_params(self):
+    #     self.assertIsNotNone(gen.categories_request(None, None).get("error"))
+
+    def test_valid_params(self):
+        response = gen.categories_request("en_US", "GB")
+        self.assertIsNone(response.get("error"))
+        self.assertIsNotNone(response.get("items"))
```

Comment:
categories_request function should return a dictionary if the input arguments are valid, and should raise an error (perhaps valueError?) when something is wrong. In the case of a valid argument, it might be useful to check if there is an "items" key.